### PR TITLE
encapsulate signer

### PR DESCRIPTION
**Pull request by richardpringle**
Depends on https://github.com/ava-labs/coreth/pull/693. But https://github.com/ava-labs/coreth/pull/693 also depends on this. Not sure about merge order... 

## Why this should be merged

Currently, signing happens all over the place and should definitely be abstracted away, especially when dealing with sensitive data like a private key. 

Once this abstraction lands, we could further enhance security by using secure hardware and signing outside of the avalanchego application and potentially integrate platforms like cubist. 

## How this works

Encapsulate signing such that the secret-key does not leave our `bls` utility package. 
The main changes are in this file: `utils/crypto/bls/secret.go`

## How this was tested

CI

## Need to be documented in RELEASES.md?

I don't think so?

### CODE REVIEW

File: go.mod
Comment (richardpringle) [inline go.mod:194]: Not sure what the process is here

Comment (richardpringle) [inline go.mod:194]: Not sure what the process is here

File: network/config.go
Comment (richardpringle) [inline network/config.go:131]: I assume this means the `BLSKey` property gets serialized as the string, `"-"`?

File: chains/manager.go
Comment (richardpringle) [inline chains/manager.go:189]: Do we want to change the property name here too?

File: config/config.go
Comment (richardpringle) [inline config/config.go:697]: I'll have to change this eventually to fully abstract the signer. I'm open to ideas on how best to do this

File: network/network_test.go
Comment (richardpringle) [inline network/network_test.go:178]: 🤔 If we want to change all the `*Key` names (variables or properties), I would recommend/request that I do that in a follow-up to limit the scope of both changes.

**Comment by yacovm** [comment]
> Once this abstraction lands, we could further enhance security by using secure hardware and signing outside of the avalanchego application and potentially integrate platforms like cubist.


Do they support BLS? Most hardware security modules only support ECDSA / ed25519...

**Comment by richardpringle** [comment]
> > Once this abstraction lands, we could further enhance security by using secure hardware and signing outside of the avalanchego application and potentially integrate platforms like cubist.
> 
> Do they support BLS? Most hardware security modules only support ECDSA / ed25519...

Yeah, they specifically have support for Ethereum validators

File: utils/crypto/bls/secret.go
Comment (richardpringle) [inline utils/crypto/bls/secret.go:26]: I should get rid of this type

File: network/config.go
Comment (joshua-kim) [reply network/config.go:131]: IIRC this actually omits it from serialization, i.e it's not serialized or deserialized. This is done because we use the `Config` struct to also hold fields and not just config values.... so our hack is to ignore serialization for non-config values (which is weird IMO).

Comment (richardpringle) [reply network/config.go:131]: Makes sense. 

Yeah, I generally prefer to split up raw config data vs actual dependent services/structs, but 🤷. I've seen this pattern before too.

File: utils/crypto/bls/secret.go
Comment (joshua-kim) [inline utils/crypto/bls/secret.go:1]: We should probably rename this to something like `signer.go`

File: utils/crypto/bls/secret.go
Comment (joshua-kim) [inline utils/crypto/bls/secret.go:44]: nit: export `signer` and return it as a concrete type here so that the calling code can decide whether to use it as an interface/as a concrete type. We'll have to figure out a fix/good naming scheme since it'll collide w/ the interface name as it currently is though. We could also as an alternative argue that the interface is premature and just use the struct.

File: utils/crypto/bls/secret.go
Comment (richardpringle) [reply utils/crypto/bls/secret.go:1]: My preference would be to make that change in a follow up if you're cool with that? Just trying to make the absolute minimal possible changes here. 

However, it's effectively impossible to mess that up, so if you feel strongly, I can do that before merging this.

File: utils/crypto/bls/secret.go
Comment (richardpringle) [reply utils/crypto/bls/secret.go:44]: 🤔  I don't think the interface is premature though, I don't plan on adding any functionality for the other implementations. What I'm trying to get is encapsulation so that you can't access the secret-key outside of this package. I think want to make a `LocalSigner`, then some "cloud-based" signers (there are a couple of options here that will each have their own concrete types).   

Having it generic from day 1 minimizes the impact of converting what's here to a `LocalSigner` (which would also be used in all tests), and other signer types would be created based on config values on application start. 

Thoughts?

File: utils/crypto/bls/secret.go
Comment (joshua-kim) [reply utils/crypto/bls/secret.go:1]: I think it's fine to do that in whatever PR we do the var renames in

File: snow/context.go
Comment (richardpringle) [reply snow/context.go:103]: It's used differently than `snowtest.Context`. I think it makes sense to move this back to `coreth` though

**Review by joshua-kim — APPROVED** [review]
Another idea is to cache the key bytes and public key but not going to block on it.

File: go.mod
Comment (richardpringle) [inline go.mod:194]:
```suggestion
```

**Review by aaronbuchwald — COMMENTED** [review]
Could we update to a coreth rc rather than a commit before merging to master? After that, this LGTM

File: config/config.go
Comment (richardpringle) [reply config/config.go:697]: Yeah, looking at the code again, signer creation would be specific. A `LocalSigner` should actually write the key to disk before returning from a constructor. We could have an `EphemeralSigner` that's similar but doesn't save the key and is used for testing. 

With that said, I would probably create a 
```go
func initializeSigner(signerConfig) (Signer) { /* ... */ }
```
and deal with saving or not saving there. It should not be dealt with here IMO. 

This is all for a future PR though

File: go.mod
Comment (richardpringle) [reply go.mod:13]: yep

File: network/config.go
Comment (richardpringle) [reply network/config.go:131]: one day...

File: go.mod
Comment (richardpringle) [reply go.mod]: done

File: utils/crypto/bls/secret.go
Comment (StephenButtolph) [inline utils/crypto/bls/secret.go:62]:
```suggestion
func SecretKeyFromBytes(skBytes []byte) (*LocalSigner, error) {
```

File: utils/crypto/bls/secret.go
Comment (StephenButtolph) [inline utils/crypto/bls/secret.go:34]: If we adopt the suggested change on `SecretKeyFromBytes` - then this can be deleted now.
```suggestion
```

File: utils/crypto/bls/secret.go
Comment (StephenButtolph) [inline utils/crypto/bls/secret.go:32]: I feel like these functions are going to need to return `error` once we want to use them for remote signing. Are we waiting to add `error` to the return value for when that other implementation is introduced?

Comment (richardpringle) [reply utils/crypto/bls/secret.go:32]: Yeah, I don't think there's any point in pre-maturely defining the interface here. It's unclear to me at this point whether or not we'll want the specific signer implementation to handle its own errors or wether we would want the caller to handle errors (I'm missing some details on how to gracefully shutdown). 

I'm interested to hear your thoughts here, but I definitely don't think this is something that we should deal with yet.

File: utils/crypto/bls/secret.go
Comment (richardpringle) [reply utils/crypto/bls/secret.go]: done fd7f3268a81550f73917cbe36baa1ccef46a639d

File: utils/crypto/bls/secret.go
Comment (StephenButtolph) [reply utils/crypto/bls/secret.go:32]: Sounds good. I think the only complex question will be around the PoP in the networking code... Everything else should be fairly straightforward I think... But 🤷 we can deal with that when we need to I suppose.

**Review by StephenButtolph — APPROVED** [review]

**Comment by meaghanfitzgerald** [comment]
https://github.com/ava-labs/avalanchego/issues/3640 Re:
> Within the config, I'd expect us to have a switch that offers different BLS Signers ie. generate and store a new BLS key on the machine or use an external signing service.
> I'd think this is either AvalancheGo defining the options it wants to support directly here or it supports a gRPC service that dials an external signer and handles signature requests.

**Pull request merged**
